### PR TITLE
Units Page for Displaying, Creating and Editing Units

### PR DESCRIPTION
**Pull request by andrew-cline**
# Description
Adds a new page for displaying all units currently stored in the database. This page allows users to create new units and edit new units via popup modal UI elements.

## Type of change

- [ ] Note merging this changes the node modules
- [ ] Note merging this changes the database configuration.
- [ ] This change requires a documentation update

## Checklist

- [x] I have followed the [OED pull request](https://openenergydashboard.github.io/developer/pr.html) ideas
- [x] I have removed text in ( ) from the issue request

## Limitations

- Need to route adding a unit through the state. Otherwise, you need to refresh the page for new units to appear.
- When changing the type of unit, the edit functionality will fail, producing this error from the database 
 `oed-3-database-1  | 2022-04-27 02:23:01.337 UTC [59] ERROR:  duplicate key value violates unique constraint "units_type_of_unit_unit_index_key"
oed-3-database-1  | 2022-04-27 02:23:01.337 UTC [59] DETAIL:  Key (type_of_unit, unit_index)=(meter, 4) already exists.
oed-3-database-1  | 2022-04-27 02:23:01.337 UTC [59] STATEMENT:  UPDATE units SET name = 'Unit Name', identifier = 'Identifier', unit_represent = 'quantity', sec_in_rate = 3600, type_of_unit = 'meter', unit_index = 4, suffix = 'Suffix', displayable = 'none', preferred_display = false, note = 'Note' WHERE id = 22;`

### CODE REVIEW

File: .vscode/settings.json
Comment (huss) [inline .vscode/settings.json]: I don't think this edit should be part of the PR. I'm removing it.

File: src/client/app/actions/units.ts
Comment (huss) [inline src/client/app/actions/units.ts:59]: Can two units be edited at once? I'm just wondering if this is needed.

File: src/client/app/actions/units.ts
Comment (huss) [inline src/client/app/actions/units.ts:86]: Is this used anywhere? I see use of confirmEditedUnit.

File: src/client/app/components/unit/CreateUnitModalComponent.tsx
Comment (huss) [inline src/client/app/components/unit/CreateUnitModalComponent.tsx]: When I look at other props in OED, they do not put Create in front of the name. I would remove it.

File: src/client/app/components/unit/CreateUnitModalComponent.tsx
Comment (huss) [inline src/client/app/components/unit/CreateUnitModalComponent.tsx]: unitRepresent, typeOfUnit & displayable are all enums in the DB but string here. Other places in the code it uses enum and not string. If you look at src/client/app/components/RouteComponent.tsx you see "defaultLanguage: LanguageTypes" which is defined in src/client/app/types/redux/i18n.ts. I think it would be better to use the enum rather than string to define these. The enums are defined in src/client/app/types/redux/units.ts. Also, that file has UnitData and UnitEditData (from you I think) that are very similar except for the types of these. Is the string version necessary?
Note: I just looked at the Edit modal and it does use the enum so it is a model for this one.

File: src/client/app/components/unit/CreateUnitModalComponent.tsx
Comment (huss) [inline src/client/app/components/unit/CreateUnitModalComponent.tsx]: Should this be container-fluid (spelling issue)?

File: src/client/app/components/unit/CreateUnitModalComponent.tsx
Comment (huss) [inline src/client/app/components/unit/CreateUnitModalComponent.tsx]: It would be nice to get these values from the enum (see other comment for location). Same for type_of_unit and displayable.

File: src/client/app/components/unit/UnitModalEditComponent.tsx
Comment (huss) [inline src/client/app/components/unit/UnitModalEditComponent.tsx]: oldIdentifer -> oldIdentifier.
oldunitRepresentType -> oldUnitRepresentType

File: src/client/app/components/unit/UnitModalEditComponent.tsx
Comment (huss) [inline src/client/app/components/unit/UnitModalEditComponent.tsx]: Should this be handleIdentifierChange?

File: src/client/app/components/unit/UnitModalEditComponent.tsx
Comment (huss) [inline src/client/app/components/unit/UnitModalEditComponent.tsx]: All enum types should try to make menu based on enum. See src/client/app/components/admin/CreateUserComponent.tsx for an example.

File: src/client/app/components/unit/UnitViewComponent.tsx
Comment (huss) [inline src/client/app/components/unit/UnitViewComponent.tsx:123]: Could you put in a comment on why this returns null?

File: src/client/app/components/unit/UnitsDetailComponent.tsx
Comment (huss) [inline src/client/app/components/unit/UnitsDetailComponent.tsx:45]: This is a note to remind me to add the help info.

File: src/client/app/components/RouteComponent.tsx
Comment (huss) [inline src/client/app/components/RouteComponent.tsx]: CreateUnitContainter -> CreateUnitContainer

File: src/client/app/translations/data.js
Comment (huss) [inline src/client/app/translations/data.js]: The way they are now alphabetized here needs to be copied to all the langauges.

OED does not use _ in these. They all need to be replaced with . (and in the code).

My quick scan indicates not all the strings are in all the languages.

File: src/client/app/styles/unit-card-page.css
Comment (huss) [inline src/client/app/styles/unit-card-page.css]: Should these commented out lines be here?

File: src/client/app/components/unit/UnitViewComponent.tsx
Comment (huss) [inline src/client/app/components/unit/UnitViewComponent.tsx]: These strings should use the translated keys and translate.

File: src/client/app/components/unit/UnitModalEditComponent.tsx
Comment (huss) [inline src/client/app/components/unit/UnitModalEditComponent.tsx]: This should be translated.

File: src/client/app/components/unit/UnitModalEditComponent.tsx
Comment (huss) [inline src/client/app/components/unit/UnitModalEditComponent.tsx]: All of these labels should be translated.

File: src/client/app/translations/data.js
Comment (huss) [inline src/client/app/translations/data.js:42]: You also have unit.create_new_unit which is used in the code but not in all languages. Pick one.

File: src/client/app/components/unit/UnitModalEditComponent.tsx
Comment (huss) [inline src/client/app/components/unit/UnitModalEditComponent.tsx]: Should this be in a label tag?

**Review by huss — CHANGES_REQUESTED** [review]
I need to play with this some more but I wanted to get out my first review comments now:

- There are a number of comments. I did some editing but left some to be done. I would like your thoughts and to know if you are able to do the changes (as appropriate/agree).
- It would be nice if the items in the create/edit modals spanned most of the horizontal area rather than a the limited amount they do now.
- Can the note area on the edit page be expandable as it is on the create page?
- We should show the name on the edit page. It is present on the create page.
- I think it would be fine for a non-admin to see the unit page but have no create or edit buttons. I tried to quickly change it so the option stayed on the main page but it did not let it go to the unit page. If you have time to fix this then great (or someone else needs to do it later).
- I edited some files that you should look at to see if you agree including
    - package-lock.json changed the version back to 1. This is fixed.
    - Not sure why the .vscode settings were changed (actaully empty was added) so put back.
    - Formatted some files.
    - Limited note to 30 characters on main unit page so not too big.
    - Modified some comments and spelling.
    - Fix a few variable names.
    - Modified unit-car-page.css so the card is smaller so more fit on a page. The project needs to be sure it is consistent across the new pages with cards.
    - Partly fixed up data.js for translations.

**Comment by huss** [comment]
Thanks to @andrew-cline for pointing out that there was an issue in editing units. It turned out to be related to a constraint on the units table:  
UNIQUE (type_of_unit, unit_index)  
See the design document for a discussion but the decision was made to remove this constraint. Basically, the unit_index is tricky to set and is invalid during the update of units until cik is updated. The code still needs to be altered to run the cik update but the constraint was removed in a recent commit to this PR. You can check if you have this constraint by doing:  
\d units  
and you will see something similar to  
"units_type_of_unit_unit_index_key" UNIQUE CONSTRAINT, btree (type_of_unit, unit_index)  
You can then remove this constraint with  
ALTER TABLE units DROP CONSTRAINT units_type_of_unit_unit_index_key;  
Once this is done you should no longer see an error about this. If you install the database from this commit forward then you should not have any issue.

File: src/client/app/translations/data.js
Comment (JustLe56) [reply src/client/app/translations/data.js:42]: Should be fixed as of 8b4b241.

File: src/client/app/components/unit/UnitModalEditComponent.tsx
Comment (JustLe56) [reply src/client/app/components/unit/UnitModalEditComponent.tsx]: Should be fixed as of 9d59efc.

File: src/client/app/components/unit/UnitModalEditComponent.tsx
Comment (JustLe56) [reply src/client/app/components/unit/UnitModalEditComponent.tsx]: Should be fixed as of 9d59efc.

File: src/client/app/components/unit/UnitViewComponent.tsx
Comment (JustLe56) [reply src/client/app/components/unit/UnitViewComponent.tsx]: Should be fixed as of 9d59efc.

File: src/client/app/translations/data.js
Comment (JustLe56) [reply src/client/app/translations/data.js]: Should be fixed as of 9d59efc.

File: src/client/app/components/unit/CreateUnitModalComponent.tsx
Comment (andrew-cline) [reply src/client/app/components/unit/CreateUnitModalComponent.tsx]: Fixed with most recent commit.

File: src/client/app/components/unit/CreateUnitModalComponent.tsx
Comment (andrew-cline) [reply src/client/app/components/unit/CreateUnitModalComponent.tsx]: Fixed with most recent commit.

File: src/client/app/components/unit/UnitModalEditComponent.tsx
Comment (andrew-cline) [reply src/client/app/components/unit/UnitModalEditComponent.tsx]: Fixed with most recent commit.

File: src/client/app/actions/units.ts
Comment (sagarprasad63574) [reply src/client/app/actions/units.ts:59]: No you cannot edit more than one unit at a time but the function submitEditedUnits() is needed to edit a unit. It would be best to change this function to edit a single unit.

File: src/client/app/actions/units.ts
Comment (sagarprasad63574) [reply src/client/app/actions/units.ts:86]: No. It can be removed

File: src/client/app/components/unit/CreateUnitModalComponent.tsx
Comment (andrew-cline) [reply src/client/app/components/unit/CreateUnitModalComponent.tsx]: Fixed in most recent commit

File: src/client/app/components/unit/CreateUnitModalComponent.tsx
Comment (andrew-cline) [reply src/client/app/components/unit/CreateUnitModalComponent.tsx]: Fixed in most recent commit

File: src/client/app/components/unit/UnitModalEditComponent.tsx
Comment (andrew-cline) [reply src/client/app/components/unit/UnitModalEditComponent.tsx]: Fixed in most recent commit

File: src/client/app/components/unit/UnitModalEditComponent.tsx
Comment (andrew-cline) [reply src/client/app/components/unit/UnitModalEditComponent.tsx]: Fixed in most recent commit

File: src/client/app/components/unit/UnitViewComponent.tsx
Comment (andrew-cline) [reply src/client/app/components/unit/UnitViewComponent.tsx:123]: Added comment for return statement in most recent commit.

File: src/client/app/components/RouteComponent.tsx
Comment (andrew-cline) [reply src/client/app/components/RouteComponent.tsx]: Fixed in most recent commit

File: src/client/app/components/unit/UnitModalEditComponent.tsx
Comment (JustLe56) [reply src/client/app/components/unit/UnitModalEditComponent.tsx]: Fixed as of 8b4b241

File: src/client/app/styles/unit-card-page.css
Comment (andrew-cline) [reply src/client/app/styles/unit-card-page.css]: Removed those comments in most recent commit.

**Comment by huss** [comment]
Many thanks to the team for addressing all the comments made. A second review indicates they are all done and it runs correctly. I have created the item "unit page updates" on the resource generalization project for what is left to do that is outside this work. This is ready to merge but I am waiting until review some other new modal pages to make sure no other changes are needed.

Note I made changes via a commit. The team is welcome (encouraged :-) to review them to make sure they all seem okay. The commit message lists what was done.

**Comment by huss** [comment]
This PR is being held until we decide if any changes are needed to reconcile how other teams did similar work.

**Comment by huss** [comment]
React hooks have now been integrated into OED. The work in this PR has been integrated into that work and moved forward. Thus, the current plan is to close this PR and move forward with a new one.

**Pull request merged**
